### ISSUE TITLE
GPU Usage

### ISSUE THREAD

**Issue by eplab1745** [yhz-1995/cudaica_win#2]
First of all, wonderful job with restarting this project.

I have it running (Matlab 2018a, Titan Black). I installed CUDA 8.1.

However, though the ICA finishes quickly, the GPU usage according to windows is 0-1%. Is that normal? GPU memory is being used and CPU process is 20-40%

**Comment by CloudyDory** [comment]
Hi, I think Windows 10 cannot reliably report the actual GPU usage when GPU is used for cuda computing. I notice the same phenomenon on my machine. As long as your ICA decomposition gives reasonable result, it should be fine.

**Comment by jimkroger** [comment]
> Hi, I think Windows 10 cannot reliably report the actual GPU usage when GPU is used for cuda computing. I notice the same phenomenon on my machine. As long as your ICA decomposition gives reasonable result, it should be fine.

yhz, can you summarize the state of the windows implementation of CUDAICA? I've read this github site as well as Fraimono’s. Is the windows implementation running as well as Fraimono’s  does under linux? Has the reliability of the ICA results been verified by comparing with other ICAs as Fraimono did?

Thanks.
Jim

**Comment by CloudyDory** [comment]
Hi Jimkroger, thanks for your interest in CUDAICA.

**1. Is the windows implementation running as well as Fraimono’s does under linux?**
Based on the test result on my machine (Intel i7-7700HQ + NVIDIA 1050Ti), the current windows implementation runs about 3 times faster than the original Fraimono's implementation (for reason see below). I have not bench-marked the speed on Linux vs Windows on other hardware because none of our lab's computers had Linux installed at the time of development.

Three main enhancement of Windows implementation are listed here:
(1). The current Windows implementation supports doing PCA before ICA. Note that the PCA calculation and the combination of PCA weights into ICA weights are done in cudaica.m (not in the exe file). Therefore, if you use the PCA option, the resulte files generated by cudaica.exe is NOT your ICA result of the original data, but the result of PCA-ed data. The output of cudacia.m is always the ICA result of the original data.

(2). The current Windows implementation uses CUDA kernel when calculating "YU=Y*U", and "W=lrate * YU * W+W", whereas the Fraimono's implementation uses CUBLAS functions (see Figure 1 of Fraimono's 2012 paper). This is the main reason why CUDAICA on Windows now runs about 3 times faster than Linux's version. I am very surprised when I found this because Fraimono had actually written all the CUDA kernels, but somehow he commented them out and switched to CUBLAS functions in the source code.

(3). The current Windows implementation uses double-precision numbers, and (if I remembered correctly), the Fraimono's implementation use single-precision numbers. Single-precision calculation will generate NaN results when whitening the data if your data is rank deficient (see [issue 5 on Fraimono’s Github site](https://github.com/fraimondo/cudaica/issues/5)).

**2. Has the reliability of the ICA results been verified by comparing with other ICAs as Fraimono did?**
I have directly compared the ICA decomposition results between CUDAICA on Windows and EEGLAB's runica. The EEG data used are from a 63-channel BrainVision device. For the extended infomax algorithm, if you set the random number generator and block size the same on CUDAICA and runica, you will get very similar results. The small differences are due to different precision on CPU and GPU, as well as non-associativity of floating point number calculation, and also depends on whether your data is rank-deficient or not. 

For small dataset (in my test case, a 64*697500 samples data), the differences are negligible and your will get all the ICs that have the same weight, same activation, and aligned in the same order. For larger data, the differences will accumulate over steps and gives you a bit different, but largely similar results. However, if you enable PCA and make the data rank-full, the differences between runica vs CUDAICA results will be much smaller even for large sized data.

Note that CUDAICA and runica uses different block size by default. I set the block size to be the same when verifying the result because I hope to check all the formulas are implemented correctly. CUDAICA uses larger block size to train ICA weights in order to take advantage of GPU. However I don't expect systematic differences because they use the same algorithm.

I have uploaded a ppt file on Github page that shows some results when we verify the CUDAICA results "2018-09-10 CUDAICA on Windows.pptx". Please check if it can further clarify some of your questions.

**Comment by jimkroger** [comment]
Yunhui, thanks very much for the very detailed reply. I would like to ry this but I am a little nervous I may not have the programming knowleege for building and compiling. I know it's not hard but have not much experience. Maybe I can do it.

Is it necessary to recompile the code? I think it is necessary for CUDAICA to work I need to replce or modify the two files in eeglab.

I have a Dell 7820 with 12 cores, 64 g of ram, and a Quadro P4000. Does this indicate if I need to rebuild the code? It would be great to have CUDAICA working under windows so I don't have to make a dual boot machine. 

Thanks!

Jim Kroger

**Comment by CloudyDory** [comment]
Hi Jim,

It is not necessary to compile the source code. The EEGLAB_Plugin folder in this Github page contains a compiled binary (cudaica_win.exe). It works on Windows 10 and Windows 7 machines with NVIDIA GTX 750/750Ti/1050Ti/1070/1080Ti, RTX 2080, and Quadro M2000 GPU, CUDA version 9/10, and Intel Paralle Studio 2018/2019 in our lab.  Other operating systems or NVIDIA GPU are not tested because we don't have them.  But given CUDA's excellent compatibility, I don't think GPU model should be a problem as long as you have enough GPU RAM for your data. Issues are more likely to occur on the Windows part.

I would recommend you to try the compiled binary first. It is necessary to install Intel MKL and NVIDIA CUDA as they are CUDAICA's dependencies, and add the Intel MKL libraries' path to Windows' system path. Modifying or replacing the two files in EEGLAB is only necessary if you want to call CUDAICA from GUI. They are not needed if you directly call cudaica.m and manually incorporate the results into the EEG structure. 

To check the whether the dependencies are met, open Windows command line in the folder containing cudaica_win.exe, and type "cudaica_win". If you see an output telling you all the CUDA devices and help information, you should be fine. Otherwise you may not receive any output. If you don't see any output, double-click on "cudaica_win.exe" and Windows should pop-out a message telling you what dll are missing.

In case you receive any errors when running CUDAICA in Matlab, always check the output of cudaica.exe in Matlab command window, even if Matlab tells you which line fails in cudaica.m.  If you see the iteration information of each step like in runica, then cudaica_win.exe should run fine. If you see a list of CUDA devices, configuration file information, but no step iteration information, the most likely cudaica_win.exe has silently failed in the background. If the latter one is the case, try a small dataset (such as the example data in EEGLAB). If a small dataset runs fine, then your GPU memory might be too small. If a small data won't run, post the error information here and we will try to find the reason.

If cudaica_win.exe or Matlab directly crashes (you receive a pop-out saying cudaica_win.exe has stopped working) after calling CUDAICA, you may need to recompile the source code. This may happen on some Windows 8 machine. To compile the source code you need to install Microsoft Visual Studio, and then reinstall the Visual Studio Integration components in Intel Parallel Studio and NVIDIA CUDA. Then double-click the "CUDAICA_Win.sln" file to open the whole project. Try compiling the code by following the help in Visual Studio. Compilation of the source code is not that complicated. There are still some settings to tweak, but it has already been done and stored in Visual Studio's project files on this Github page. Post any error information here and I will see if I can help you.

Best,
Yunhui

**Comment by winndsd** [comment]
CUDA Usage can not be monitored by Windows Task Manager. However, it is actually works when you can find GPU memory loading uprising a little in TM.
The logging information can be recorded when you try 'NVIDIA Nsight system', which is installed with CUDA Toolkit.


Debug in line 237 of cudaica.m when running EEGLAB to keep the file 'cudaica.sc' available.

Try the following command line in  'NVIDIA Nsight system', which should be modified to your own path.

E:\eeglab_git\eeglabgit\plugins\CudaICA1.0\cudaica_win.exe -f "D:\DESKTOP\cudaica.sc"

You will get the real usage percentage when running the ICA task,

I'm using Intel Core i7-9750H and NVIDIA GeFore RTX 2060, and I got 53.8% usage at most in my report.


![Snipaste_2022-03-03_02-05-21](https://user-images.githubusercontent.com/28735316/156424710-ee3f374b-074a-4acd-8fef-b62bd8130385.png)
